File: app/src/main/java/com/fastaccess/ui/adapter/LabelsAdapter.java
package com.fastaccess.ui.adapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.ViewGroup;

import com.fastaccess.data.dao.LabelModel;
import com.fastaccess.ui.adapter.viewholder.LabelsViewHolder;
import com.fastaccess.ui.widgets.recyclerview.BaseRecyclerAdapter;
import com.fastaccess.ui.widgets.recyclerview.BaseViewHolder;

import java.util.List;

/**
 * Created by Kosh on 11 Nov 2016, 2:07 PM
 */

public class LabelsAdapter extends BaseRecyclerAdapter<LabelModel, LabelsViewHolder, BaseViewHolder
        .OnItemClickListener<LabelModel>> {

    public interface OnSelectLabel {
        boolean isLabelSelected(LabelModel labelModel);

        void onToggleSelection(LabelModel labelModel, boolean select);
    }

    @Nullable private OnSelectLabel onSelectLabel;

    public LabelsAdapter(@NonNull List<LabelModel> eventsModels, @Nullable OnSelectLabel onSelectLabel) {
        super(eventsModels);
        this.onSelectLabel = onSelectLabel;
    }

    @Override protected LabelsViewHolder viewHolder(ViewGroup parent, int viewType) {
        return LabelsViewHolder.newInstance(parent, onSelectLabel, this);
    }

    @Override protected void onBindView(LabelsViewHolder holder, int position) {
        holder.bind(getItem(position));
    }
}

File: app/src/main/java/com/fastaccess/github/ui/adapter/LabelsAdapter.kt
package com.fastaccess.github.ui.adapter

import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.fastaccess.data.model.parcelable.LabelModel
import com.fastaccess.github.ui.adapter.viewholder.LabelViewHolder

/**
 * Created by Kosh on 20.01.19.
 */
class LabelsAdapter : ListAdapter<LabelModel, LabelViewHolder>(DIFF_CALLBACK) {

    val selection = hashSetOf<LabelModel>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LabelViewHolder = LabelViewHolder(parent).apply {
        itemView.let { view ->
            view.setOnClickListener {
                val position = adapterPosition
                val item = getItem(position)
                if (!selection.contains(item)) {
                    selection.add(item)
                } else {
                    selection.remove(item)
                }
                notifyItemChanged(position)
            }
        }
    }

    override fun onBindViewHolder(holder: LabelViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item, selection.contains(item))
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<LabelModel?>() {
            override fun areItemsTheSame(oldItem: LabelModel, newItem: LabelModel): Boolean = oldItem.name == newItem.name
            override fun areContentsTheSame(oldItem: LabelModel, newItem: LabelModel): Boolean = oldItem == newItem
        }
    }
}
